04.ls/ls.rb

```diff
@@ -1,33 +1,166 @@
 #!/usr/bin/env ruby
 # frozen_string_literal: true
 
+require 'optparse'
+require 'etc'
+require 'date'
+
 COLUMNS = 3
 BLANK = 2
 
+FILE_TYPE = {
+  '01' => 'p',
+  '02' => 'c',
+  '06' => 'b',
+  '10' => '-',
+  '12' => 'l',
+  '14' => 's',
+  '40' => 'd'
+}.freeze
+
+PERMISSION_MODE = {
+  '000' => '---',
+  '001' => '--x',
+  '010' => '-w-',
+  '011' => '-wx',
+  '100' => 'r--',
+  '101' => 'r-x',
+  '110' => 'rw-',
+  '111' => 'rwx'
+}.freeze
+
 def main
   default_files = Dir.glob('*')
-  files = ARGV.include?('-r') ? default_files.reverse : default_files
-  output(files)
+  if ARGV.include?('-l')
+    LongOption.new.output(default_files)
+  else
+    Default.new.output(default_files)
+  end
 end
 
-def build_file_grid(files)
-  remainder = files.size % COLUMNS
-  padding_count = (COLUMNS - remainder) % COLUMNS
-  padded_files = files + Array.new(padding_count, '')
-  row_count = padded_files.size / COLUMNS
-  file_grid = padded_files.each_slice(row_count).to_a.transpose
-  column_width = files.map(&:length).max
+class LongOption
+  def output(default_files)
+    files = default_files
+    puts "total #{FileInformation.new.total(files)}"
+
+    files.each do |file|
+      puts [
+        FileInformation.new.permission(file),
+        FileInformation.new.hard_link(file).to_s.rjust(column_width(default_files)[:hard_link]),
+        FileInformation.new.owner_name(file),
+        FileInformation.new.group_name(file),
+        FileInformation.new.file_size(file).to_s.rjust(column_width(default_files)[:file_size]),
+        FileInformation.new.last_modified_month(file).to_s.rjust(column_width(default_files)[:last_modified_month]),
+        FileInformation.new.last_modified_day(file),
+        FileInformation.new.last_modified_hour_minute(file),
+        FileInformation.new.file_name(file)
+      ].join(' ')
+    end
+  end
+
+  private
 
-  [file_grid, column_width]
+  def column_width(default_files)
+    {
+      hard_link: default_files.map { |file| File.stat(file).nlink.to_s.length }.max,
+      file_size: default_files.map { |file| File.stat(file).size.to_s.length }.max,
+      last_modified_month: default_files.map { |file| "#{File.mtime(file).month}月".to_s.length }.max
+    }
+  end
 end
 
-def output(files)
-  file_grid, column_width = build_file_grid(files)
-  file_grid.each do |files|
-    files.each do |file|
-      print file.to_s.ljust(column_width + BLANK, ' ')
+class FileInformation
+  def total(default_files)
+    total = default_files.map { |default_file| File.stat(default_file).blocks }
+    total.flatten.sum
+  end
+
+  def permission(file)
+    [
+      file_type(file),
+      owner(file),
+      group(file),
+      other_group(file)
+    ].join('')
```

Comment:
```suggestion
    ].join
```

デフォルト値と同じなので省略してもいいかも

Reply:
省略する形に修正しました！
ご確認よろしくお願い致します！